lib/eventq/eventq_aws/aws_queue_client.rb

```diff
@@ -3,71 +3,84 @@ module Amazon
     class QueueClient
 
       def initialize(options = {})
-        if options.has_key?(:aws_key)
-          Aws.config[:credentials] = Aws::Credentials.new(options[:aws_key], options[:aws_secret])
-        end
-
-        if !options.has_key?(:aws_account_number)
-          raise ':aws_account_number option must be specified.'.freeze
-        end
-
-        @aws_account = options[:aws_account_number]
+        invalid_keys = options.keys - [:sns_keep_alive_timeout, :sns_continue_timeout ]
+        raise(OptionParser::InvalidOption, invalid_keys) unless invalid_keys.empty?
 
         @sns_keep_alive_timeout = options[:sns_keep_alive_timeout] || 30
         @sns_continue_timeout = options[:sns_continue_timeout] || 15
-
-        if options.has_key?(:aws_region)
-          @aws_region = options[:aws_region]
-          Aws.config[:region] = @aws_region
-        else
-          @aws_region = Aws.config[:region]
-        end
       end
 
       # Returns the AWS SQS Client
       def sqs
-        @sqs ||= Aws::SQS::Client.new
+        @sqs ||= sqs_client
       end
 
       # Returns the AWS SNS Client
       def sns
-        @sns ||= Aws::SNS::Client.new(
-          http_idle_timeout: @sns_keep_alive_timeout,
-          http_continue_timeout: @sns_continue_timeout
-        )
-      end
-
-      def get_topic_arn(event_type)
-        _event_type = EventQ.create_event_type(event_type)
-        return "arn:aws:sns:#{@aws_region}:#{@aws_account}:#{aws_safe_name(_event_type)}"
+        @sns ||= sns_client
       end
 
       def get_queue_arn(queue)
-        _queue_name = EventQ.create_queue_name(queue.name)
-        return "arn:aws:sqs:#{@aws_region}:#{@aws_account}:#{aws_safe_name(_queue_name)}"
+        _, arn = get_queue_url(queue)
+        arn
       end
 
       def create_topic_arn(event_type)
         _event_type = EventQ.create_event_type(event_type)
         response = sns.create_topic(name: aws_safe_name(_event_type))
         return response.topic_arn
       end
+      alias :get_topic_arn :create_topic_arn
 
-      # Returns the URL of the queue. The queue will be created when it does
+      # Returns the URL and ARN of the queue. The queue will be created when it does
       #
       # @param queue [EventQ::Queue]
-      def get_queue_url(queue)
+      # @return url,arn [Array]
+      def get_queue_url_arn(queue)
         _queue_name = EventQ.create_queue_name(queue.name)
         response= sqs.get_queue_url(
-                                     queue_name: aws_safe_name(_queue_name),
-                                     queue_owner_aws_account_id: @aws_account,
-                                   )
-        return response.queue_url
+          queue_name: aws_safe_name(_queue_name)
+        )
+        result = sqs.get_queue_attributes({queue_url: response.queue_url, attribute_names: ['QueueArn']})
+
+        return response.queue_url, result.attributes['QueueArn']
```

Comment:
Ideal place to add the url and arn to an in memory cache to avoid making multiple calls to aws from the same process for the same queue.